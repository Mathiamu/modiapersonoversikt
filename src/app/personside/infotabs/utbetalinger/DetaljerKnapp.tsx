import * as React from 'react';
import styled from 'styled-components';
import { theme } from '../../../../styles/personOversiktTheme';
import NavFrontendChevron from 'nav-frontend-chevron';

interface Props {
    onClick: (event: React.MouseEvent<HTMLButtonElement>) => void;
    open: boolean;
}

const KnappWrapper = styled.button`
  border: none;
  padding: 0;
  border-radius: 0.5em;
  cursor: pointer;
  background-color: transparent;
  &:hover {
    background-color: ${theme.color.hoverLink};
  }
  &:focus {
    ${theme.focus}
  }
`;

function DetaljerKnapp(props: Props) {
    return (
        <KnappWrapper onClick={props.onClick} aria-expanded={props.open} aria-label="Detaljer">
            <NavFrontendChevron type={props.open ? 'opp' : 'ned'}/>
        </KnappWrapper>
    );
}

export default DetaljerKnapp;
