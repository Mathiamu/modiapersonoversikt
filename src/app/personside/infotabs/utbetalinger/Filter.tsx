import * as React from 'react';
import styled from 'styled-components';
import { Checkbox, Radio } from 'nav-frontend-skjema';
import { Element } from 'nav-frontend-typografi';
import EtikettMini from '../../../../components/EtikettMini';
import NavDatovelger from 'nav-datovelger';
import { Feilmelding } from '../../../../utils/Feilmelding';

export interface FilterState {
    periode: Periode;
    utbetaltTil: UtbetaltTilValg;
    ytelse: Ytelse;
}

interface FraTil {
    fra?: Date;
    til?: Date;
}

interface Periode {
    radioValg: PeriodeValg;
    egendefinertPeriode: FraTil;
}

interface UtbetaltTilValg {
    bruker: boolean;
    annenMottaker: boolean;
}

interface Ytelse {
    alleYtelser: boolean;
}

interface Props {
    onChange: (change: Partial<FilterState>) => void;
    filterState: FilterState;
}

const FiltreringsPanel = styled.nav`
  > *:first-child {
    margin-top: 2em;
  }
`;

const InputPanel = styled.form`
  margin-top: 1em;
  margin-bottom: 1em;
  > *:not(:last-child) {
    margin-bottom: 1em;
  }
`;

const Opacity = styled.span`
  opacity: .5;
`;

export enum PeriodeValg {
    SISTE_30_DAGER = 'Siste 30 dager',
    INNEVÆRENDE_ÅR = 'Inneværende år',
    I_FJOR = 'I fjor',
    EGENDEFINERT = 'Egendefinert'
}

function onRadioChange(props: Props, key: PeriodeValg) {
    props.onChange({
        periode: {
            ...props.filterState.periode,
            radioValg: key
        }
    });
}

function onDatoChange(props: Props, dato: Partial<FraTil>) {
    props.onChange({
        periode: {
            ...props.filterState.periode,
            egendefinertPeriode: {
                ...props.filterState.periode.egendefinertPeriode,
                ...dato
            }
        }
    });
}

function onUtbetaltTilChange(props: Props, change: Partial<UtbetaltTilValg>) {
    props.onChange({
        utbetaltTil: {
            ...props.filterState.utbetaltTil,
            ...change
        }
    });
}

function onYtelseChange(props: Props, change: Partial<Ytelse>) {
    props.onChange({
        ytelse: {
            ...props.filterState.ytelse,
            ...change
        }
    });
}

function egendefinertDatoInputs(props: Props) {

    const fra = props.filterState.periode.egendefinertPeriode.fra;
    const til = props.filterState.periode.egendefinertPeriode.til;
    const periodeFeilmelding = fra && til && fra > til ?
        <Feilmelding feil={{feilmelding: 'Fra-dato kan ikke være senere enn til-dato'}}/> : null;

    return (
        <>
            <label htmlFor="utbetalinger-datovelger-fra">Fra:</label>
            <NavDatovelger
                dato={props.filterState.periode.egendefinertPeriode.fra}
                id={'utbetalinger-datovelger-fra'}
                avgrensninger={{maksDato: new Date()}}
                onChange={dato => onDatoChange(props, {fra: dato})}
            />
            <label htmlFor="utbetalinger-datovelger-til">Til:</label>
            <NavDatovelger
                dato={props.filterState.periode.egendefinertPeriode.til}
                id={'utbetalinger-datovelger-til'}
                avgrensninger={{maksDato: new Date()}}
                onChange={dato => onDatoChange(props, {til: dato})}
            />
            {periodeFeilmelding}
        </>
    );
}

function Filtrering(props: Props) {

    const radios = Object.keys(PeriodeValg).map(key => {
        const label = PeriodeValg[key];
        const checked = props.filterState.periode.radioValg === label;
        return (
            <Radio
                label={label}
                key={label}
                checked={checked}
                onChange={() => onRadioChange(props, PeriodeValg[key])}
                name="FiltreringsvalgGruppe"
            />);
    });

    return (
        <FiltreringsPanel>
            <Element>Filtrering</Element>

            <InputPanel>
                <EtikettMini><Opacity>Velg periode</Opacity></EtikettMini>
                {radios}
                {props.filterState.periode.radioValg === PeriodeValg.EGENDEFINERT && egendefinertDatoInputs(props)}
            </InputPanel>

            <InputPanel>
                <EtikettMini><Opacity>Utbetaling til</Opacity></EtikettMini>
                <Checkbox
                    label="Bruker"
                    checked={props.filterState.utbetaltTil.bruker}
                    onChange={() => onUtbetaltTilChange(props, {bruker: !props.filterState.utbetaltTil.bruker})}
                />
                <Checkbox
                    label="Annen mottaker"
                    checked={props.filterState.utbetaltTil.annenMottaker}
                    onChange={() => onUtbetaltTilChange(
                        props,
                        {annenMottaker: !props.filterState.utbetaltTil.annenMottaker}
                    )}
                />
            </InputPanel>

            <InputPanel>
                <EtikettMini><Opacity>Velg ytelse</Opacity></EtikettMini>
                <Checkbox
                    label="Alle"
                    checked={props.filterState.ytelse.alleYtelser}
                    onChange={() => onYtelseChange(props, {alleYtelser: !props.filterState.ytelse.alleYtelser})}
                />
            </InputPanel>

        </FiltreringsPanel>
    );
}

export default Filtrering;
