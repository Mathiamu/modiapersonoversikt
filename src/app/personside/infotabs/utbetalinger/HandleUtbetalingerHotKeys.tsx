import * as React from 'react';
import { Utbetaling, Ytelse } from '../../../../models/utbetalinger';
import { flatMapYtelser } from './utils/utbetalingerUtils';
import { Dispatch } from 'redux';
import {
    setEkspanderYtelse,
    setNyYtelseIFokus
} from '../../../../redux/utbetalinger/utbetalingerStateReducer';
import { AppState } from '../../../../redux/reducers';
import { connect } from 'react-redux';

interface OwnProps {
    children: React.ReactNode;
    utbetalinger: Utbetaling[];
}

interface DispatchProps {
    settYtelseIFokus: (ytelse: Ytelse) => void;
    ekspanderYtelse: (ytelse: Ytelse, ekspander: boolean) => void;
}

interface StateProps {
    ytelseIFokus: Ytelse | null;
    ekspanderteYtelser: Ytelse[];
}

type Props = DispatchProps & OwnProps & StateProps;

class HandleUtbetalingerHotKeys extends React.Component<Props> {

    constructor(props: Props) {
        super(props);
        this.handleKeyDown = this.handleKeyDown.bind(this);
    }

    handleKeyDown(event: React.KeyboardEvent) {
        const eventTargetIsButton = event.target instanceof HTMLButtonElement;
        if (eventTargetIsButton) {
            return;
        }
        this.handlePilknapper(event);
        this.handleEnter(event);
    }

    handlePilknapper(event: React.KeyboardEvent) {
        if (event.key === 'ArrowDown') {
            this.props.settYtelseIFokus(this.finnNesteYtelse());
        } else if (event.key === 'ArrowUp') {
            this.props.settYtelseIFokus(this.finnForrigeYtelse());
        }
    }

    finnNesteYtelse() {
        const ytelser: Ytelse[] = flatMapYtelser(this.props.utbetalinger);
        const currentIndex = this.props.ytelseIFokus ? ytelser.indexOf(this.props.ytelseIFokus) : -1;
        return ytelser[currentIndex + 1] || ytelser[0];
    }

    finnForrigeYtelse() {
        const ytelser: Ytelse[] = flatMapYtelser(this.props.utbetalinger);
        const currentIndex = this.props.ytelseIFokus ? ytelser.indexOf(this.props.ytelseIFokus) : ytelser.length;
        return ytelser[currentIndex - 1] || ytelser[ytelser.length - 1];
    }

    handleEnter(event: React.KeyboardEvent) {
        if (event.key === 'Enter' && !event.repeat) {
            this.toggleEkspanderYtelseIFokus();
        }
    }

    toggleEkspanderYtelseIFokus() {
        const ytelseIFokus = this.props.ytelseIFokus;
        if (!ytelseIFokus) {
            return;
        }
        const erEkspandert = this.props.ekspanderteYtelser.includes(ytelseIFokus);
        this.props.ekspanderYtelse(ytelseIFokus, !erEkspandert);
    }

    render() {
        return (
            <div onKeyDown={this.handleKeyDown}>
                {this.props.children}
            </div>
        );
    }
}

function mapDispatchToProps(dispatch: Dispatch<{}>): DispatchProps {
    return {
        settYtelseIFokus: ytelse => dispatch(setNyYtelseIFokus(ytelse)),
        ekspanderYtelse: (ytelse: Ytelse, ekspander: boolean) => dispatch(setEkspanderYtelse(ytelse, ekspander))
    };
}

function mapStateToProps(state: AppState): StateProps {
    return {
        ytelseIFokus: state.utbetalinger.ytelseIFokus,
        ekspanderteYtelser: state.utbetalinger.ekspanderteYtelser
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(HandleUtbetalingerHotKeys);
