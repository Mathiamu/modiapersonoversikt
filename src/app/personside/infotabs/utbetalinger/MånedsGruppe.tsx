import * as React from 'react';
import styled from 'styled-components';
import theme from '../../../../styles/personOversiktTheme';
import { ArrayGroup } from '../../../../utils/groupArray';
import { Normaltekst } from 'nav-frontend-typografi';
import { Bold, Uppercase } from '../../../../components/common-styled-components';
import { getGjeldendeDatoForUtbetaling } from './utils/utbetalingerUtils';
import UtbetalingsKomponent from './utbetaling/Utbetaling';
import { Utbetaling } from '../../../../models/utbetalinger';

interface MånedsgruppeProps {
    gruppe: ArrayGroup<Utbetaling>;
}

const MånedGruppeStyle = styled.li`
  > *:first-child {
    background-color: ${theme.color.kategori};
    padding: .2rem ${theme.margin.px20};
  }
  ol {
    padding: 0;
    margin: 0;
  }
  ol > *:not(:first-child) {
    border-top: ${theme.border.skille};
  }
`;

function Månedsgruppe({gruppe}: MånedsgruppeProps) {
    const utbetalingsKomponenter = gruppe.array.map(utbetaling => (
        <UtbetalingsKomponent
            key={getGjeldendeDatoForUtbetaling(utbetaling) + utbetaling.nettobeløp}
            utbetaling={utbetaling}
        />
    ));
    return (
        <MånedGruppeStyle>
            <Normaltekst tag={'h3'}><Bold><Uppercase>{gruppe.category}</Uppercase></Bold></Normaltekst>
            <ol>
                {utbetalingsKomponenter}
            </ol>
        </MånedGruppeStyle>
    );
}

export default Månedsgruppe;
