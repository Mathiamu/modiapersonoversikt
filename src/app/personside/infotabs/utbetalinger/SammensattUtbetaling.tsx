import * as React from 'react';
import {
    datoVerbose,
    formaterNOK,
    getGjeldendeDatoForUtbetaling,
    getNettoSumYtelser,
    periodeStringFromYtelse
} from './utbetalingerUtils';
import { UtbetalingStyle } from './Utbetaling';
import styled from 'styled-components';
import { Bold, SpaceBetween, UndertekstGrå, UndertekstGråBold } from '../../../../components/common-styled-components';
import PrintKnapp from '../../../../components/PrintKnapp';
import { UndertekstBold } from 'nav-frontend-typografi';
import DetaljerKnapp from './DetaljerKnapp';
import UtbetalingsDetaljer from './UtbetalingsDetaljer';
import { Utbetaling, Ytelse } from '../../../../models/utbetalinger';

export interface EnkeltYtelseProps {
    ytelse: Ytelse;
    konto: string | undefined;
    melding: string | undefined;
}

interface EnkeltYtelseState {
    visDetaljer: boolean;
}

const EnkeltYtelseStyle = styled<{ focus: boolean }, 'li'>('li')`
  transition: 0.3s;
  ${props => props.focus && 'background-color: rgba(0, 0, 0, 0.03);'}
`;

const PadLeft = styled.div`
  margin-left: .5rem;
`;

class EnkeltYtelse extends React.Component<EnkeltYtelseProps, EnkeltYtelseState> {

    constructor(props: EnkeltYtelseProps) {
        super(props);
        this.state = {
            visDetaljer: false
        };
        this.toggleVisDetaljer = this.toggleVisDetaljer.bind(this);
    }

    toggleVisDetaljer() {
        this.setState({
            visDetaljer: !this.state.visDetaljer
        });
    }

    render() {
        const ytelse = this.props.ytelse;
        const periode = periodeStringFromYtelse(ytelse);
        const detaljer = this.state.visDetaljer &&
            <UtbetalingsDetaljer ytelse={ytelse} konto={this.props.konto} melding={this.props.melding}/>;

        return (
            <EnkeltYtelseStyle focus={this.state.visDetaljer}>
                <SpaceBetween>
                    <UndertekstGråBold>
                        {ytelse.type}
                    </UndertekstGråBold>
                    <SpaceBetween>
                        <UndertekstGråBold>
                            {formaterNOK(ytelse.nettobeløp)}
                        </UndertekstGråBold>
                        <PadLeft/>
                        <DetaljerKnapp onClick={this.toggleVisDetaljer} open={this.state.visDetaljer}/>
                    </SpaceBetween>
                </SpaceBetween>
                <UndertekstGrå>
                    {periode}
                </UndertekstGrå>
                {detaljer}
            </EnkeltYtelseStyle>
        );
    }
}

interface Props {
    utbetaling: Utbetaling;
}

const YtelsesListe = styled.ul`
  padding-left: .5rem;
  list-style: none;
  > * {
    margin-top: .2rem;
  }
`;

function SammensattUtbetaling(props: Props) {
    const utbetaling = props.utbetaling;
    if (!utbetaling.ytelser) {
        return <>'Manglende data i utbetaling.'</>;
    }

    const dato = datoVerbose(getGjeldendeDatoForUtbetaling(utbetaling)).sammensatt;
    const sum = formaterNOK(getNettoSumYtelser(utbetaling.ytelser));
    const forfallsInfo = utbetaling.forfallsdato && !utbetaling.utbetalingsdato
        ? `Forfallsdato: ${dato}` : '';
    const ytelsesListe = utbetaling.ytelser.map((ytelse, index) => (
        <EnkeltYtelse
            ytelse={ytelse}
            konto={utbetaling.konto}
            melding={utbetaling.melding}
            key={index}
        />
    ));

    return (
        <UtbetalingStyle>
            <SpaceBetween>
                <UndertekstGrå>
                    {dato} / <Bold>{utbetaling.status}</Bold>
                </UndertekstGrå>
                <PrintKnapp onClick={() => alert('ikke implementert')}/>
            </SpaceBetween>
            <SpaceBetween>
                <UndertekstBold tag={'h4'}>{'Diverse ytelser'}</UndertekstBold>
                <UndertekstBold>{sum}</UndertekstBold>
            </SpaceBetween>
            <SpaceBetween>
                <UndertekstGrå>Utbetaling til: {utbetaling.utbetaltTil}</UndertekstGrå>
                <UndertekstGrå>{forfallsInfo}</UndertekstGrå>
            </SpaceBetween>
            <YtelsesListe>
                {ytelsesListe}
            </YtelsesListe>
        </UtbetalingStyle>
    );
}

export default SammensattUtbetaling;
