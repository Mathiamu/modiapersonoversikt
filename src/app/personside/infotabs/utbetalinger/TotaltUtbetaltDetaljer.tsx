import * as React from 'react';
import styled from 'styled-components';
import { TotaltUtbetaltProps } from './TotaltUtbetalt';
import { Undertekst } from 'nav-frontend-typografi';
import { Utbetaling, Ytelse, Ytelseskomponent } from '../../../../models/utbetalinger';
import { groupArray } from '../../../../utils/groupArray';
import theme from '../../../../styles/personOversiktTheme';
import {
    filtrerBortUtbetalingerSomIkkeErUtbetalt,
    formaterNOK,
    getBruttoSumYtelser,
    getNettoSumYtelser,
    getPeriodeFromYtelser,
    getTrekkSumYtelser, getTypeFromYtelse, reduceUtbetlingerTilYtelser
} from './utbetalingerUtils';
import { formaterDato } from '../../../../utils/dateUtils';
import { Fragment } from 'react';
import ErrorBoundary from '../../../../components/ErrorBoundary';

const Wrapper = styled.div`
  margin: 1.5rem 0 .5rem;
  th {
    font-weight: bold;
  }
  tr {
    display: flex;
    flex-flow: row wrap;
    position: relative; /* IE11 */
    > *:first-child {
      flex-grow: 1;
      display: block; /* IE11 */
    }
    .sumBrutto, .sumNetto, .sumTrekk {
      position: absolute; /* IE11 */
    }
    .sumBrutto {
      right: 12rem; /* IE11 */
    }
    .sumTrekk {
      right: 6rem; /* IE11 */
    }
    .sumNetto {
      right: 0; /* IE11 */
      font-weight: bold;
    }
    .periodeForYtelse, .ytelseDetaljer {
      flex-basis: 100%;
      opacity: .7;
    }
    .periodeForYtelse {
      display: block; /* IE11 */
      text-align: left;
    }
    .visually-hidden {
      ${theme.visuallyHidden}
    }
    .ytelseDetaljer {
      display: block; /* IE11 */
      flex-grow: 1;
      text-align: right;
      margin-right: 12rem;
      dl {
        display: flex;
        flex-flow: row wrap;
        dt {
          flex-basis: 50%;
          flex-grow: 1;
        }
        dd {
          padding-left: 1rem;
          flex-basis: 7rem;
        }
      }
    }
  }
  tbody tr {
    margin-top: .5rem;
    > *:nth-last-child(n+3) {
        padding-top: .5rem;
        border-top: 2px solid ${theme.color.bakgrunn};
    }
  }
`;

function getAlleYtelsesKomponenterFraYtelser(ytelser: Ytelse[]) {
    return ytelser.reduce(
        (acc: Ytelseskomponent[], ytelse: Ytelse) => {
            if (!ytelse.ytelseskomponentListe) {
                throw new Error('Ytelse mangler ytelseskomponenter');
            }
            return [...acc, ...ytelse.ytelseskomponentListe];
        },
        []
    );
}

function getYtelsesKomponentSammendragListe(ytelser: Ytelse[]) {
    const alleYtelsesKomponenter = getAlleYtelsesKomponenterFraYtelser(ytelser);
    const ytelsesKomponenterGruppertPåType = groupArray(
        alleYtelsesKomponenter,
        ytelseskomponent => ytelseskomponent.ytelseskomponenttype
    );
    const listeKomponenter = ytelsesKomponenterGruppertPåType.map(gruppe => {
        const sum = gruppe.array.reduce((acc, ytelsesKomponent) => acc + ytelsesKomponent.ytelseskomponentbeløp, 0);
        return (
            <Fragment key={gruppe.category}>
                <dt>{gruppe.category}</dt>
                <dd>{formaterNOK(sum)}</dd>
            </Fragment>
        );
    });
    return <dl>{listeKomponenter}</dl>;
}

function getAlleUtbetalteYtelserFraUtbetalinger(utbetalinger: Utbetaling[]) {
    const utbetalteUtbetalinger = utbetalinger.filter(filtrerBortUtbetalingerSomIkkeErUtbetalt);
    return reduceUtbetlingerTilYtelser(utbetalteUtbetalinger);
}

function getYtelserSammendrag(utbetalinger: Utbetaling[]) {
    const alleUtbetalteYtelser: Ytelse[] = getAlleUtbetalteYtelserFraUtbetalinger(utbetalinger);
    const ytelserGruppertPåTema = groupArray(
        alleUtbetalteYtelser,
        getTypeFromYtelse
    );
    const ytelsesSammendrag = ytelserGruppertPåTema.map(gruppe => {
        const ytelser = gruppe.array;
        const ytelsesType = gruppe.category;
        const periode = getPeriodeFromYtelser(ytelser);
        const ytelsesKomponentSammendragListe = getYtelsesKomponentSammendragListe(ytelser);
        return (
            <tr key={ytelsesType}>
                <th>{ytelsesType}</th>
                <td className="sumBrutto">{formaterNOK(getBruttoSumYtelser(ytelser))}</td>
                <td className="sumTrekk">{formaterNOK(getTrekkSumYtelser(ytelser))}</td>
                <td className="sumNetto">{formaterNOK(getNettoSumYtelser(ytelser))}</td>
                <td className="periodeForYtelse">{formaterDato(periode.fra)} - {formaterDato(periode.til)}</td>
                <td className="ytelseDetaljer">{ytelsesKomponentSammendragListe}</td>
            </tr>
        );
    });
    return ytelsesSammendrag;
}

function TotaltUtbetaltDetaljer(props: TotaltUtbetaltProps) {
    const ytelserSammendrag = getYtelserSammendrag(props.utbetalinger);
    return (
        <ErrorBoundary>
            <Wrapper>
                <Undertekst tag="span">
                    <h2>Sammendrag</h2>
                    <table>
                        <thead>
                        <tr>
                            <th>Ytelse</th>
                            <th className="sumBrutto">Brutto</th>
                            <th className="sumTrekk">Trekk</th>
                            <th className="sumNetto">Utbetalt</th>
                            <th className="visually-hidden">Periode</th>
                            <th className="visually-hidden">Detaljer</th>
                        </tr>
                        </thead>
                        <tbody>
                        {ytelserSammendrag}
                        </tbody>
                    </table>
                </Undertekst>
            </Wrapper>
        </ErrorBoundary>
    );
}

export default TotaltUtbetaltDetaljer;
