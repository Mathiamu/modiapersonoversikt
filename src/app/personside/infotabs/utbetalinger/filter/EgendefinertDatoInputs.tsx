import * as React from 'react';
import { UtbetalingFilterState, FraTilDato } from '../../../../../redux/utbetalinger/types';
import { formaterDato, formaterTilISO8601Date } from '../../../../../utils/stringFormatting';
import { DatovelgerAvgrensninger } from 'nav-datovelger';
import moment from 'moment';
import Datovelger from 'nav-datovelger/dist/datovelger/Datovelger';
import { Feilmelding } from '../../../../../utils/Feilmelding';
import { tidligsteTilgjengeligeDatoUtbetalingerRestkonto } from '../../../../../redux/restReducers/utbetalinger';
import { Periode } from '../../../../../models/periode';
import { isValidDate } from '../../../../../utils/dateUtils';

interface Props {
    filter: UtbetalingFilterState;
    updateFilter: (change: Partial<UtbetalingFilterState>) => void;
}

function onDatoChange(props: Props, dato: Partial<Periode>) {
    const newPeriode: FraTilDato = {
        fra: dato.fra ? moment(dato.fra).toDate() : props.filter.periode.egendefinertPeriode.fra,
        til: dato.til ? moment(dato.til).toDate() : props.filter.periode.egendefinertPeriode.til
    };
    props.updateFilter({
        periode: {
            ...props.filter.periode,
            egendefinertPeriode: newPeriode
        }
    });
}

function getDatoFeilmelding(fra: Date, til: Date) {
    if (fra > til) {
        return <Feilmelding feil={{ feilmelding: 'Fra-dato kan ikke være senere enn til-dato' }} />;
    }
    if (til > new Date()) {
        return <Feilmelding feil={{ feilmelding: 'Du kan ikke velge dato frem i tid' }} />;
    }
    if (fra < tidligsteTilgjengeligeDatoUtbetalingerRestkonto) {
        return (
            <Feilmelding
                feil={{
                    feilmelding: `Du kan ikke velge en dato før ${formaterDato(
                        tidligsteTilgjengeligeDatoUtbetalingerRestkonto
                    )}`
                }}
            />
        );
    }
    if (!isValidDate(fra) || !isValidDate(til)) {
        return (
            <Feilmelding
                feil={{
                    feilmelding: 'Du må velge gyldige datoer'
                }}
            />
        );
    }
    return null;
}

function EgendefinertDatoInputs(props: Props) {
    const fra = props.filter.periode.egendefinertPeriode.fra;
    const til = props.filter.periode.egendefinertPeriode.til;
    const periodeFeilmelding = getDatoFeilmelding(fra, til);
    const avgrensninger: DatovelgerAvgrensninger = {
        minDato: formaterTilISO8601Date(tidligsteTilgjengeligeDatoUtbetalingerRestkonto),
        maksDato: formaterTilISO8601Date(new Date())
    };

    return (
        <>
            <label htmlFor="utbetalinger-datovelger-fra">Fra:</label>
            <Datovelger
                input={{ id: 'utbetalinger-datovelger-fra', name: 'Fra dato' }}
                visÅrVelger={true}
                valgtDato={formaterTilISO8601Date(props.filter.periode.egendefinertPeriode.fra)}
                onChange={dato => onDatoChange(props, { fra: dato })}
                id="utbetalinger-datovelger-fra"
                avgrensninger={avgrensninger}
            />
            <label htmlFor="utbetalinger-datovelger-til">Til:</label>
            <Datovelger
                input={{ id: 'utbetalinger-datovelger-til', name: 'Til dato' }}
                visÅrVelger={true}
                valgtDato={formaterTilISO8601Date(props.filter.periode.egendefinertPeriode.til)}
                onChange={dato => onDatoChange(props, { til: dato })}
                id="utbetalinger-datovelger-til"
                avgrensninger={avgrensninger}
            />
            {periodeFeilmelding}
        </>
    );
}

export default EgendefinertDatoInputs;
