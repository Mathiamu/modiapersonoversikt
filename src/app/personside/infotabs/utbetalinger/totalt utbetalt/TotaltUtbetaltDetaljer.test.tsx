import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { statiskMockUtbetaling } from '../../../../../mock/statiskMockUtbetaling';
import TotaltUtbetaltDetaljer from './TotaltUtbetaltDetaljer';
import { FilterState, PeriodeValg } from '../filter/Filter';
import { aremark } from '../../../../../mock/person/aremark';
import { Utbetaling } from '../../../../../models/utbetalinger';
import { ReactNode } from 'react';

Date.now = jest.fn(() => new Date()); // for Ã¥ motvirke Date.now() mock i setupTests.ts

// Mock react collapse sin UnmountClosed
jest.mock('react-collapse', () => {
    return {
        UnmountClosed: (props: {children: ReactNode}) => props.children
    };
});

const filterState: FilterState = {
    periode: {
        radioValg: PeriodeValg.SISTE_30_DAGER,
        egendefinertPeriode: {
            fra: new Date(0),
            til: new Date(0)
        }
    },
    utbetaltTil: [aremark.navn.sammensatt],
    ytelser: ['Sykepenger']
};

const mockUtbetalingUtenUtbetaltDato: Utbetaling = {
    ...statiskMockUtbetaling,
    utbetalingsdato: null
};

const mockUtbetalingReturnertForSaksbehandling: Utbetaling = {
    ...statiskMockUtbetaling,
    status: 'Returnert til NAV for saksbehandling'
};

test('Filtrerer bort utbetalinger som ikke skal medregnes og viser totalt-utbetalt-detaljer riktig', () => {
    const visittkortheader = renderer.create(
        <TotaltUtbetaltDetaljer
            visDetaljer={true}
            toggleVisDetaljer={() => null}
            filter={filterState}
            utbetalinger={[
                statiskMockUtbetaling,
                statiskMockUtbetaling,
                mockUtbetalingUtenUtbetaltDato,
                mockUtbetalingReturnertForSaksbehandling
            ]}
        />
    );

    const json = visittkortheader.toJSON();
    expect(json).toMatchSnapshot();
});
