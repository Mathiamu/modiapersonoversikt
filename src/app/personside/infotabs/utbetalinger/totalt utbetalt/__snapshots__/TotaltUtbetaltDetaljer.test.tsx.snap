// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filtrerer bort utbetalinger som ikke skal medregnes og viser totalt-utbetalt-detaljer riktig 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c0 {
  -webkit-transition: .5s;
  transition: .5s;
  border-radius: 0.25rem;
  background-color: #f4f4f4;
  box-shadow: 0 0 0 1px #59514b;
  padding: 1.25rem 0.625rem;
  margin: 1.25rem 0;
}

.c3 {
  border: none;
  padding: .1rem .2rem;
  margin-top: 1rem;
  border-radius: .35rem;
  cursor: pointer;
  background-color: transparent;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #0067c5;
}

.c3:focus {
  outline: none;
  box-shadow: 0 0 0 0.1875rem #FFBD66;
}

.c3:hover {
  opacity: 0.8;
}

.c4 {
  margin-right: .5rem;
}

.c1 th:not(:first-child) {
  font-weight: normal;
}

.c1 th {
  text-transform: uppercase;
}

.c1 tr {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  position: relative;
}

.c1 tr > *:first-child {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  display: block;
}

.c1 tr .sumBrutto,
.c1 tr .sumNetto,
.c1 tr .sumTrekk {
  position: absolute;
}

.c1 tr .sumBrutto {
  right: 12rem;
}

.c1 tr .sumTrekk {
  right: 6rem;
}

.c1 tr .sumNetto {
  right: 0;
}

.c1 tr td.sumNetto {
  font-weight: bold;
}

.c1 tr .periodeForYtelse,
.c1 tr .ytelseDetaljer {
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
}

.c1 tr .periodeForYtelse {
  display: block;
  text-align: left;
}

.c1 tr .ytelseDetaljer {
  display: block;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  text-align: right;
  margin-right: 12rem;
}

.c1 tr .ytelseDetaljer dl {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
}

.c1 tr .ytelseDetaljer dl dt {
  -webkit-flex-basis: 50%;
  -ms-flex-preferred-size: 50%;
  flex-basis: 50%;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c1 tr .ytelseDetaljer dl dd {
  padding-left: 1rem;
  -webkit-flex-basis: 7rem;
  -ms-flex-preferred-size: 7rem;
  flex-basis: 7rem;
}

.c1 tbody tr {
  margin-top: .5rem;
}

.c1 tbody tr > *:nth-last-child(n+3) {
  padding-top: .5rem;
  border-top: 2px solid #efefef;
}

<div
  className="c0"
  open={true}
>
  <aside
    className="c1"
  >
    <span
      className="typo-undertekst"
    >
      <h2>
        Sammendrag
      </h2>
      <table>
        <thead>
          <tr>
            <th>
              Ytelse
            </th>
            <th
              className="sumBrutto"
            >
              Brutto
            </th>
            <th
              className="sumTrekk"
            >
              Trekk
            </th>
            <th
              className="sumNetto"
            >
              Utbetalt
            </th>
            <th
              className="visually-hidden"
            >
              Periode
            </th>
            <th
              className="visually-hidden"
            >
              Detaljer
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th>
              Sykepenger 2017
            </th>
            <td
              className="sumBrutto"
            >
              26,740.00
            </td>
            <td
              className="sumTrekk"
            >
              -3,968.00
            </td>
            <td
              className="sumNetto"
            >
              22,772.00
            </td>
            <td
              className="periodeForYtelse"
            >
              09.11.2016
               - 
              31.10.2017
            </td>
            <td
              className="ytelseDetaljer"
            >
              <dl>
                <dt>
                  Mouse
                </dt>
                <dd>
                  7,720.00
                </dd>
                <dt>
                  Computer
                </dt>
                <dd>
                  7,860.00
                </dd>
                <dt>
                  Bacon
                </dt>
                <dd>
                  11,160.00
                </dd>
              </dl>
            </td>
          </tr>
        </tbody>
      </table>
    </span>
  </aside>
  <div
    className="c2"
  >
    <button
      aria-expanded={true}
      className="c3"
      onClick={[Function]}
      open={true}
    >
      <span
        className="c4"
      >
        <span
          className="typo-normal"
        >
          Skjul
           detaljer
        </span>
      </span>
      <i
        className="nav-frontend-chevron chevronboks chevron--opp"
      />
    </button>
  </div>
</div>
`;
